openlibrary/plugins/openlibrary/js/jquery.repeat.js

```diff
@@ -74,6 +80,8 @@ export default function($){
             $.extend(data, options.vars || {});
 
             newid = elems._this.attr("id") + "--" + index;
+            // increment the index to avoid situations where more than one element have same
+            index++;
```

Comment:
I'm not crazy about this solution. This will ensure that the indices are unique, but still has the issue that they are not actually indices. They're more like identifiers. Indices should be consecutive. Instead of this, `onRemove` should be updated so that all the sibling elements of the removed element have their indices decremented.

I'm not sure if having holes in the indices will cause issues, but the word "index" is misleading in the way it is used right now, and could lead future developers astray.

Reply:
The issue is that the IDs need to be unique for when the form posts. The alternative to this is to rewrite the entire DOM and update all the IDs of the children on every add or remove - so if there are 20 rows and 1 id is removed then 20 nodes need rewriting. This is expensive and unnecessary.

Is the issue only the variable name index? I am happy to rename this but I'm not sure what to - essentially it needs to make sure the ID gets uniquely generated on each addition - this is the bug - right now it doesn't and has likely been broken for many years.

Reply:
We're _assuming_ they only have to be unique IDs, but the name "index" does not imply that. Just in case some other code is assuming the index to actually be an index, unless we want to research all uses of this (which would non-trivial because these names are getting used in backend code as well), I think it's safer, faster, and clearer to update the dom.

The performance hit is super negligible. Removing an element form an array is always an $O(n)$ operation, and that's fine. Also, since we know n to be almost always less than 20, that's doubly fine. Also, since this is an event triggered by user input that won't happen automatically--triply fine :P This feels like premature optimization.

Reply:
> We're assuming they only have to be unique IDs, but the name "index" does not imply that.

The IDs are elements inside a FORM element and have to be unique to be treated differently when they are posted, otherwise duplicate values will be ignored as they are. I'm not seeing what's wrong with incrementing an index that's used to generate a new index (it's better than 'identifier-' + Math.random(), as this ensures that the new element always has a unique ID which is what we're trying to solve. What is your problem with this? Just the name? Would using the variable name new_identifier_index be sufficient?

Rewriting the DOM (e.g. modifying siblings) as well as modifying things that do not need to be modified will require wiring up new events, updating references to the new IDs, possibly some changes to the backend and I suspect more complications that I've not thought about.

I'd prefer to keep this solution as simple as possible as it is currently written.

Reply:
> We're assuming they **only** have to be unique IDs, but the name "index" does not imply that.

I 100% agree with you that they have to be unique :P "My problem with this" is that they're not consecutive, which could cause issues later on. I've explained why; do you understand?

You're saying that you don't want to make them consecutive because you don't want to update the DOM because you think it would complicate the code unnecessarily and take a lot of devtime. Is that right?

I initially thought that this was only being used to populate the element's id, but it looks like the index is also getting passed to the template file, so there's no way this method can know everywhere `index` is being used unless it parses the template HTML. That's why it would be difficult to update all sibling indices. Once the rebasing is done, I'll test to see if it has any issues with holes in the indices, but I think it should be ok.

Let's rename the template value from `index` to `rowId`; note this would also need to be updated in the `html` files; a search for `{{index}}` should find them. Sound good?

Reply:
The index key is shared between the validate function Renaming the template variable will require changes to the python side of thing and the template code.

I don't understand how either of those things work (yet), so I am reluctant to change them (they will likely require backend changes as well).

Can this be the subject of a new issue?

Reply:
👍 Sounds good; I'll create a new issue!